src/libs/protoboard/blackboard_manager.h

```diff
@@ -0,0 +1,331 @@
+	friend AbstractProtobufSender;
+
+	ProtobufThead *                         message_handler_;
+	fawkes::ProtobufPeerInterface *         peer_iface_;
+	pb_conversion_map                       bb_receiving_interfaces_;
+	fawkes::BlackBoardOnMessageWaker *      on_message_waker_;
+	unsigned int                            next_peer_idx_;
+	std::unique_ptr<AbstractProtobufSender> pb_sender_;
+
+	void add_peer(fawkes::ProtobufPeerInterface *iface, long peer_id);
+
+	template <class MessageT, class InterfaceT>
+	bool handle_message_type(InterfaceT *iface);
+
+	template <class InterfaceT>
+	struct on_interface
+	{
+		InterfaceT *       iface;
+		BlackboardManager *manager;
+
+		on_interface(InterfaceT *iface, BlackboardManager *manager) : iface(iface), manager(manager)
+		{
+		}
+
+		template <class MessageT>
+		bool
+		handle_msg_types()
+		{
+			return manager->handle_message_type<MessageT>(iface);
+		}
+
+		// This template is disabled if MessageTs is {} to resolve ambiguity
+		template <class MessageT1, class... MessageTs>
+		typename std::enable_if<(sizeof...(MessageTs) > 0), bool>::type
+		handle_msg_types()
+		{
+			return handle_msg_types<MessageTs...>() || handle_msg_types<MessageT1>();
+		}
+	};
+};
+
+template <class... IfaceManagerTs>
+ProtobufSender<IfaceManagerTs...>::ProtobufSender(BlackboardManager *bb_mgr)
+: AbstractProtobufSender(bb_mgr)
+{
+}
+
+template <class... IfaceManagerTs>
+void
+ProtobufSender<IfaceManagerTs...>::init()
+{
+	boost::fusion::for_each(bb_sending_interfaces_, [this](auto &iface_mgr) {
+		iface_mgr.init(this->bb_manager->get_blackboard(), this->bb_manager);
+	});
+}
+
+template <class... IfaceManagerTs>
+void
+ProtobufSender<IfaceManagerTs...>::finalize()
+{
+	boost::fusion::for_each(bb_sending_interfaces_,
+	                        [this](auto &iface_mgr) { iface_mgr.finalize(); });
+}
+
+template <class IfaceT, class MessageT>
+bool
+AbstractProtobufSender::handle_messages::
+operator()(const bb_iface_manager<IfaceT, type_list<MessageT>> &pair) const
+{
+	return manager->handle_message_type<MessageT>(pair.interface());
+}
+
+template <class IfaceT, class MessageT1, class... MessageTs>
+bool
+AbstractProtobufSender::handle_messages::
+operator()(const bb_iface_manager<IfaceT, type_list<MessageT1, MessageTs...>> &iface_mgr) const
+{
+	return BlackboardManager::on_interface<IfaceT>{iface_mgr.interface(), manager}
+	         .template handle_msg_types<MessageTs...>()
+	       || manager->handle_message_type<MessageT1>(iface_mgr.interface());
+}
+
+template <class MessageT, class InterfaceT>
+bool
+BlackboardManager::handle_message_type(InterfaceT *iface)
+{
+	if (!iface->msgq_empty()) {
+		bool rv = false;
+		while (MessageT *msg = iface->msgq_first_safe(msg)) {
+			try {
+				handle_message(iface, msg);
+			} catch (std::exception &e) {
+				logger->log_error(
+				  name(), "Exception handling %s on %s: %s", msg->type(), iface->uid(), e.what());
+			}
+			iface->msgq_pop();
+			rv = true;
+		}
+		iface->write();
```

Comment:
Why write here? You didn't update any data, did you?

Reply:
We don't know that. `handle_message()` is a user-implemented function that may well update interface fields. Current implementations don't do that AFAIK, but it's possible, so I think better be safe than sorry.

Reply:
I see, I missed that. Why not move `iface-write()` into the `try` block?

Reply:
Right, that makes sense